src/FakeItEasy.Analyzer.Shared/ArgumentConstraintOutsideCallSpecAnalyzer.cs

```diff
@@ -0,0 +1,144 @@
+                "FakeItEasy.A.CallTo",
+                "FakeItEasy.A.CallTo`1",
+                "FakeItEasy.Fake`1.CallsTo",
+                "FakeItEasy.Fake`1.CallsTo`1",
+                "FakeItEasy.Configuration.IPropertySetterAnyValueConfiguration`1.To");
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(DiagnosticDefinitions.ArgumentConstraintOutsideCallSpec);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.RegisterSyntaxNodeAction(AnalyzeConstraint, SyntaxKind.SimpleMemberAccessExpression);
+        }
+
+        private static void AnalyzeConstraint(SyntaxNodeAnalysisContext context)
+        {
+            var memberAccess = context.Node as MemberAccessExpressionSyntax;
+            if (memberAccess == null)
+            {
+                return;
+            }
+
+            var propertySymbol = SymbolHelpers.GetAccessedPropertySymbol(memberAccess, context);
+            if (propertySymbol == null)
+            {
+                return;
+            }
+
+            var propertyFullName =
+                string.Concat(propertySymbol.ContainingType.GetFullName(), ".", propertySymbol.Name);
+
+            if (!ArgumentConstraintProperties.Contains(propertyFullName))
+            {
+                return;
+            }
+
+            if (!IsInArgumentToMethodThatSupportsArgumentConstraints(context.Node, context))
+            {
+                var descriptor = DiagnosticDefinitions.ArgumentConstraintOutsideCallSpec;
+                var completeNode = FindCompleteNode(context.Node);
+                var diagnostic = Diagnostic.Create(descriptor, completeNode.GetLocation(), completeNode.ToString());
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+
+        private static SyntaxNode FindCompleteNode(SyntaxNode node)
+        {
+            while (node.Parent.Kind() == SyntaxKind.SimpleMemberAccessExpression || node.Parent.Kind() == SyntaxKind.InvocationExpression)
+            {
+                node = node.Parent;
+            }
+
+            return node;
+        }
+
+        private static bool IsInArgumentToMethodThatSupportsArgumentConstraints(SyntaxNode node, SyntaxNodeAnalysisContext context)
+        {
+            while (node != null)
+            {
+                switch (node.Kind())
+                {
+#if CSHARP
+                    case SyntaxKind.PropertyDeclaration:
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.ConstructorDeclaration:
+#elif VISUAL_BASIC
+                    case SyntaxKind.GetAccessorBlock:
+                    case SyntaxKind.SetAccessorBlock:
+                    case SyntaxKind.SubBlock:
+                    case SyntaxKind.ConstructorBlock:
+#endif

```

Comment:
I just want to stop searching up the ancestor nodes when I reach the method level (whether the "method" is a method, property accessor or constructor. Unfortunately the node kinds are not the same for VB and C#. I found those using the Roslyn syntax tree visualizer in LinqPad.